pkg/rpc/server/server.go

```diff
@@ -270,17 +270,39 @@ func (s *Server) handleHTTPRequest(w http.ResponseWriter, httpRequest *http.Requ
 		return
 	}
 
-	err := req.DecodeData(httpRequest.Body)
+	err := in.DecodeData(httpRequest.Body)
```

Comment:
And that's where you can create a batch `if err != nil` and try decoding that one.

Reply:
The problem is that we still have to decode the request body into some neutral structure (json.RawMessage), because we cannot read it twice. And perform standard decoding after that. Having this code somewhere in a separate marshaller (i.e. `(r *Request) UnmarshalJSON`) allows us to avoid code duplication between `http` and `ws` clients. Are you sure that we should perform unmarshalling for different structures right here?

Reply:
Let me take a look at the current version, you're right in that unified unmarshaler can be better for `ws`.